### PR TITLE
KAFKA-10134: Enable heartbeat during PrepareRebalance and Depend On State For Poll Timeout

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Split the consumer coordinator's `REBALANCING` state into `PREPARING_REBALANCE ` and `COMPLETING_REBALANCE`. The first is when the join group request is sent, and the second is after the join group response is received. During the first state we should still not send hb since it shares the same socket with the join group request and the group coordinator has disabled timeout, however when we transit to the second state we should start sending hb in case leader's `assign` takes long time. This is also for fixing KAFKA-10122.

2. When deciding `coordinator#timeToNextPoll`, do not count in timeToNextHeartbeat if the state is in `UNJOINED` or `PREPARING_REBALANCE` since we would disable hb and hence its timer would not be updated.

3. On the broker side, allow hb received during `PREPARING_REBALANCE`, return NONE error code instead of REBALANCE_IN_PROGRESS. However on client side, we still need to ignore REBALANCE_IN_PROGRESS if state is COMPLETING_REBALANCE in case it is talking to an old versioned broker.

4. Piggy-backing a log4j improvement on the broker coordinator for triggering rebalance reason, as I found it a bit blurred during the investigation. Also subsumed https://github.com/apache/kafka/pull/9038 with log4j improvements.

The tricky part for allowing hb during `COMPLETING_REBALANCE` is in two parts: 1) before the sync-group response is received, a hb response may have reset the generation; also after the sync-group response but before the callback is triggered, a hb response can still reset the generation, we need to handle both cases by checking the generation / state. 2) with the hb thread enabled, the sync-group request may be sent by the hb thread even if the caller thread did not call `poll` yet.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:531]: This is a piggy-backed cleanup: we call resetGenerationXXX in the join/sync-group handler func and hence should not re-call it again here.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:647]: Minor log4j improvement.

**Review by guozhangwang — COMMENTED** [review]
ping @hachikuji @vvcephei @abbccdda  for reviews.

The open question is, whether we should avoid setting the state to UNJOINED but stay with REBALANCING if join/sync response has retriable error. E.g. if Join returned require-member-id but time is out, we would return from poll while the state is in UNJOINED.

I've thought about allowing the heartbeat to also be sent during UNJOINED, but later I felt it is an overkill since we would send hb even before the first poll call. LMK WDYT.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:647]: There might be slight performance gain if we just say "Attempt to join group and receive member id required error." instead of passing in the error.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (abbccdda) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:290]: Just to confirm, this file only has logging changes right?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: We should still handle fatal exception IMHO, such as FencedInstanceIdException

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java
Comment (abbccdda) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: remove `thread`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java
Comment (abbccdda) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:642]: Why do we need to respond?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Do we still need this private function then?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:531]: Comment here as no better place: on L485 we have this logic:
```
if (joinFuture == null) {
            // fence off the heartbeat thread explicitly so that it cannot interfere with the join group.
            // Note that this must come after the call to onJoinPrepare since we must be able to continue
            // sending heartbeats if that callback takes some time.
            disableHeartbeatThread();
```
As we are ensuring the heartbeat thread working during rebalance, will this case be a bit dangerous for heartbeat disabling? Maybe we could also do a check of member status here to decide whether to disable.

**Review by abbccdda — COMMENTED** [review]
Thanks for the PR, left a couple of comments

**Comment by abbccdda** [comment]
Checkstyle failure:

```> Task :clients:checkstyleTest FAILED
[ant:checkstyle] [ERROR] /Users/boyang.chen/code/secondary/kafka-2/clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:57:8: Unused import - java.time.Duration. [UnusedImports]```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (vvcephei) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I can see it either way. It seems like this PR is about sending the heartbeats _optimistically_ during rebalance, so there doesn't seem to really be any harm in ignoring the response for now. If we ignore the errors, then everything should still work, as the JoinGroup or SyncGroup response will tell us that we've been fenced next time we poll.

It seems like the advantage of handling the error here is that we can potentially rejoin just a tiny bit sooner by not having to wait for the JoinGroup or SyncGroup response. But it's not clear to me that it's actually ok not to handle those responses, so then we would also need to make sure the response handling logic can detect that the response has already been invalidated if we've sent a new JoinGroup request in the mean time.

This definitely has the potential to decrease the MTTR, but I'm wondering if we should take on the complexity right now, or consider it as a follow-on optimization.

**Review by vvcephei — APPROVED** [review]
Hey, @guozhangwang , thanks for this improvement!

One high level question, is this really MINOR? It seems like it's for https://issues.apache.org/jira/browse/KAFKA-10122

The change itself looks good to me. Can we be sure to run the broker compatibility system test before merging?

**Review by andrewchoi5 — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:531]: I saw this comment too, how is it that the heartbeat can interfere with the join group? Also, don't we need to remove the `disableHeartbeatThread` call that Boyang pointed to as well?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: I agree with @vvcephei , we should aim to keep this fix as simple as possible. The reason for this PR is that members are getting kicked from the group unnecessarily, not that members who are actually supposed to be fenced don't rejoin fast enough.
If we fix KAFKA-10122 and people start complaining that members fenced during a rebalance take too long to find out, then we can revisit this optimization. But if KAFKA-10122 is fixed then it seems like the occurrence of members dropping out during a rebalance will go way down anyway, so we'd be solving a vanishingly small problem

**Comment by ableegoldman** [comment]
Hey @guozhangwang , any progress on this?

**Comment by guozhangwang** [comment]
@ableegoldman I'm re-targeting this against https://issues.apache.org/jira/browse/KAFKA-10134 now as I found the root cause of that ticket is actually because we do not send (and hence reset timer of) heartbeat during rebalance, hopefully I will have this PR in a final-reviewable manner for merge.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Yup, I will just inline this then.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:531]: Yes I agree, I think we should just let the heartbeat thread access the state itself and then based on that decide whether or not to send heartbeats, I will update this logic.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: My thoughts were that, when we are in rebalancing then the purpose of heartbeat is only to keep the consumer alive at the broker side, not to take any instructions. But I think it should be handled case-by-case, I will try to refactor this piece a bit as well.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java
Comment (guozhangwang) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:642]: Actually we do not need to :)

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (guozhangwang) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:290]: Yes, it only contains logging changes. But I will make some non logging changes later and will mark it explicitly.

**Comment by guozhangwang** [comment]
@ableegoldman @vvcephei @hachikuji Please take a look at the updated description on the top.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:336]: This is the major fix 2) in description.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:489]: We do not need to explicitly disable heartbeat thread since when the state is transited to PREPARING_REBALANCE, the thread would disable itself in the next iteration.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:500]: Moved this into sync-group handler for readability.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:920]: We do not need this check any more since when we are only resetting generation if we see illegal generation or unknown member id, and in either case we should no longer heartbeat

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:1330]: This is the major fix 1).

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:1102]: This is now a redundant test.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:646]: This is the only logical change as 3) in the description. All others are logging changes.

**Review by guozhangwang — COMMENTED** [review]
Ready for reviews now.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (ableegoldman) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:642]: Honestly, if we have to keep handling the `REBALANCE_IN_PROGRESS` error on the client side anyway, then maybe it's best to keep things simple and just continue to send this in the response. Otherwise it just seems like asking for trouble if we have to consider different possible responses depending on which version of the broker the client is talking to.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (ableegoldman) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:646]: Wait, so before this the coordinator wouldn't complete the current heartbeat? Doesn't that mean that heartbeating is pointless until the rebalance completes? Obviously that doesn't line up with my observations since members were clearly getting kicked from the group before the rebalance had completed, so I must be missing something here

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Can we still specify that we failed during the JoinGroup? eg `Rebalance failed on JoinGroup with {}` or something

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:500]: I assume you mean the JoinGroup response handler 🙂

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:531]: Did you mean to say `recordRebalanceFailure` or is this comment just out of date after the latest changes?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:500]: Well I should say part of that (the enabling of the heartbeat thread) is in JoinGroup response handler, while the rest (update metrics, etc) is in SyncGroup response handler.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (guozhangwang) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:646]: It would return the error code before: that is because it does not expect clients to send heartbeat before sending sync-group requests. Now it is not the case any more.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (guozhangwang) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:642]: I had a discussion with @hachikuji about this. I think logically it should not return `REBALANCE_IN_PROGRESS` and clients in the future should update its handling logic too, maybe after some releases where we can break client-broker compatibility.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: The reason I changed it is exactly that it may not always due to join-group :) If sync-group failed, this could also be triggered.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:531]: Previously this function has two lines: update the state and record sensors. Now that the first is called in the caller, this function becomes a one-liner and hence not worthy anymore so I in-lined it.

**Review by guozhangwang — COMMENTED** [review]
@ableegoldman I think I have replied all your comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (vvcephei) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Since the state can also be set from the heartbeat thread, do you think it would be a good idea to also get a "stateSnapshot" inside the synchronized block at L439 so that the state and generation are consistent wrt each other?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (vvcephei) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Different question: can we report the exception as the "cause", rather than just getting the toString of it?
```suggestion
                log.info("Rebalance failed.", exception);
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (vvcephei) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:463]: Should we also reset the generation here? With the new condition above, we may now enter this block if generation is _not_ `NO_GENERATION`. I'm not sure if we want to have the generation set to some value but state set to `UNJOINED` and `rejoinNeeded := true`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (vvcephei) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:658]: We don't need to reset the state here anymore?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java
Comment (vvcephei) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Heartbeat.java]: Should this be inside `isTraceEnabled()` to avoid computing `remainingMs()` in the case that trace logging isn't on?

```suggestion
        if (log.isTraceEnabled()) {
            log.trace("Sending heartbeat request with {}ms remaining on timer", heartbeatTimer.remainingMs());
        }
```

(also below)

**Review by vvcephei — COMMENTED** [review]
Hey @guozhangwang , thanks for the PR!

Just a few observations. I'm not totally sure about them, though.

Thanks,
-John

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:463]: That's a good question. I just thought about this and I think I can change the caller of `resetGeneration` (which is the only place that hb thread can reset the generation) and move the `state = MemberState.UNJOINED;` into the callee to make sure that they are always changed together.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Ack, will do.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:658]: Yes, this is redundant since we are raising this error and `resetStateAndRejoin()` would still be executed at the handler anyways.

**Review by vvcephei — APPROVED** [review]
Thanks, @guozhangwang ! This addressed all my questions.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.6

**Pull request merged**
